File: 03-grid-games/00-tic-tac-toe/tictactoe-py/tic_tac_toe_game.py
class TicTacToeGame:
    def __init__(self):
        self.reset_game()

    def reset_game(self):
        self.winner = None
        self.is_game_over = False

        self.turn = 0
        self.players = ["X", "O"]

        self.board = [
            [None, None, None],
            [None, None, None],
            [None, None, None],
        ]

    def get_current_player(self):
        index = self.turn % len(self.players)
        player = self.players[index]
        return player

    def is_valid_move(self, row, col):
        if row < 0 or col < 0 or row > 2 or col > 2:
            return False

        if self.board[row][col] is not None:
            return False

        return True

    def make_move(self, row, col):
        if self.is_game_over:
            return

        if not self.is_valid_move(row, col):
            return

        # mark the board with the current player X or O
        player = self.get_current_player()
        self.board[row][col] = player

        # check to see if the move ended the game
        if self.has_winner():
            self.is_game_over = True
            self.winner = player
        elif self.turn == 8:
            self.is_game_over = True
            self.winner = "tie"
        else:
            self.turn += 1

    def check_three(self, row1, col1, row2, col2, row3, col3):
        board = self.board

        isNotNull = board[row1][col1] is not None
        oneMatchesTwo = board[row1][col1] is board[row2][col2]
        twoMatchesThree = board[row2][col2] is board[row3][col3]
        threeInARow = oneMatchesTwo and twoMatchesThree

        if isNotNull and threeInARow:
            return True
        return False

    def has_winner(self):
        lines = [
            # horizontal lines
            [0, 0, 0, 1, 0, 2],
            [1, 0, 1, 1, 1, 2],
            [2, 0, 2, 1, 2, 2],

            # vertical lines
            [0, 0, 1, 0, 2, 0],
            [0, 1, 1, 1, 2, 1],
            [0, 2, 1, 2, 2, 2],

            # top-left to bottom-right
            [0, 0, 1, 1, 2, 2],

            # bottom-left to top-right
            [2, 0, 1, 1, 0, 2]
        ]

        for line in lines:
            if self.check_three(line[0], line[1], line[2], line[3], line[4], line[5]):
                return True
        return False

File: 03-grid-games/00-tic-tac-toe/tictactoe-js/tic_tac_toe_game.js
export default class TicTacToeGame {
  constructor() {
    this.resetGame()
  }

  resetGame() {
    this.winner = null
    this.isGameOVer = false

    this.turn = 0
    this.players = ["X", "O"]

    this.board = [
      [null, null, null],
      [null, null, null],
      [null, null, null],
    ]
  }

  getCurrentPlayer() {
    const index = this.turn % this.players.length
    const player = this.players[index]
    return player
  }

  isValidMove(row, col) {
    if (row < 0 || col < 0 || row > 2 || col > 2) {
      return false
    }

    if (this.board[row][col] !== null) {
      return false
    }

    return true
  }

  makeMove(row, col) {
    if (this.isGameOver) {
      return
    }

    if (!this.isValidMove(row, col)) {
      return
    }

    // mark the board with the current player X or O
    const player = this.getCurrentPlayer()
    this.board[row][col] = player

    // check to see if the move ended the game
    if (this.hasWinner()) {
      this.isGameOVer = true
      this.winner = player
    } else if (this.turn == 8) {
      this.isGameOver = true
      this.winner = "tie"
    } else {
      this.turn += 1
    }
  }

  checkThree(row1, col1, row2, col2, row3, col3) {
    const isNotNull = this.board[row1][col1] !== null
    const oneMatchesTwo = this.board[row1][col1] === this.board[row2][col2]
    const twoMatchesThree = this.board[row2][col2] === this.board[row3][col3]
    const threeInARow = oneMatchesTwo && twoMatchesThree

    if (isNotNull && threeInARow) {
      return true
    }
    return false
  }

  hasWinner() {
    const lines = [
      // horizontal lines
      [0, 0, 0, 1, 0, 2],
      [1, 0, 1, 1, 1, 2],
      [2, 0, 2, 1, 2, 2],

      // vertical lines
      [0, 0, 1, 0, 2, 0],
      [0, 1, 1, 1, 2, 1],
      [0, 2, 1, 2, 2, 2],

      // top - left to bottom - right
      [0, 0, 1, 1, 2, 2],

      // bottom - left to top - right
      [2, 0, 1, 1, 0, 2]
    ]

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]
      if (this.checkThree(line[0], line[1], line[2], line[3], line[4], line[5])) {
        return true
      }
    }
  }
}